### PR TITLE
Fix ELM debug floating-point exceptions

### LINKED ISSUE #4820: E3SM Land Developer Tests Failing in DEBUG Mode
With current master, I ran many of the land developer tests in debug mode ( case directory is here: /lcrc/group/e3sm/ac.schwartzpd/master-debug-tests/cases) and it seems any that use the chemistry of ELM all fail.  I believe this has been an issue for a while so I can't say how many code changes this will require and not each test failed in the same place.  Almost all of them are due to the triggering of a floating point exception edit:  due to evaluation of logical expressions involving NaNs. 

A potential fix is to add `shr_infnan_isnan` consistently as that will not trigger an fpe error.  An potential downside of `shr_infnan_isnan` is I am not sure how to readily port them to use in GPU regions -- a GPU friendly check that I implemented is to use if statements like `if(.not. (x .ne. x) )` which trigger a fpe (the FireMod backtraces below).  I can make a elm wrapper function that switches between the two at compilation but will have to check how adding an extra function call impacts performance.  

Currently, many new development runs cannot be run in DEBUG mode without removal of `-fpe0` flag, and since I do not believe it is a good policy that an atm developer should have to contact an elm developer just to enable debug mode, then the `-fpe0` should be temporarily removed for elm files. And tests will need to be added to catch these issues sooner as the current ELM debug tests are clearly inadequate.
```
ERS.f09_g16.I1850ELMCN.chrysalis_intel
ERS.f09_g16.I1850GSWCNPRDCTCBC.chrysalis_intel.elm-vstrd (different traceback)
ERS.f19_f19.I20TRELMCN.chrysalis_intel
ERS.f09_g16.I1850ELMCN.chrysalis_intel.elm-bgcinterface
ERS.r05_r05.IELM.chrysalis_intel.elm-V2_ELM_MOSART_features
ERS.f19_g16.I1850CNECACNTBC.chrysalis_intel.elm-eca
ERS.f19_g16.I1850GSWCNPECACNTBC.chrysalis_intel.elm-eca_f19_g16_I1850GSWCNPECACNTBC (no traceback just seg fault)
ERS.f19_g16.I20TRGSWCNPECACNTBC.chrysalis_intel.elm-eca_f19_g16_I20TRGSWCNPECACNTBC
ERS.f19_g16.I20TRGSWCNPRDCTCBC.chrysalis_intel.elm-ctc_f19_g16_I20TRGSWCNPRDCTCBC
ERS.f19_g16.I1850CNRDCTCBC.chrysalis_intel.elm-rd (no traceback just seg fault)
ERS.f19_g16.I1850ELM.chrysalis_intel.elm-betr (betr traceback)
SMS_Ld1.hcru_hcru.I1850CRUELMCN.chrysalis_intel
SMS_Ly2_P1x1.1x1_smallvilleIA.IELMCNCROP.chrysalis_intel.elm-lulcc_sville (FireMod traceback)
SMS_Ly2_P1x1.1x1_smallvilleIA.IELMCNCROP.chrysalis_intel.elm-per_crop
SMS.r05_r05.I1850ELMCN.chrysalis_intel.elm-qian_1948
```
@rljacob @bishtgautam

### LINKED ISSUE #4946: Debug test SMS_D.ne30pg2_r05_EC30to60E2r2.BGCEXP_CNTL_CNPRDCTC_1850 fails in ELM
A debug test of SMS_D.ne30pg2_r05_EC30to60E2r2.BGCEXP_CNTL_CNPRDCTC_1850 (when PR #4944 is applied to remove mpas-o issues) fails with the following traceback:
````
2660: forrtl: error (65): floating invalid
2660: Image              PC                Routine            Line        Source
2660: libpnetcdf.so.3.0  000015555101C68C  for__signal_handl     Unknown  Unknown
2660: libpthread-2.28.s  000015554E041B20  Unknown               Unknown  Unknown
2660: e3sm.exe           00000000051EE06A  subgridavemod_mp_        1075  subgridAveMod.F90
2660: e3sm.exe           0000000006E12518  cnpbudgetmod_mp_c        1148  CNPBudgetMod.F90
2660: e3sm.exe           0000000006E11AFD  cnpbudgetmod_mp_c        1085  CNPBudgetMod.F90
2660: e3sm.exe           0000000004C90524  elm_driver_mp_elm         578  elm_driver.F90
2660: e3sm.exe           0000000004C2D3B5  lnd_comp_mct_mp_l         506  lnd_comp_mct.F90
2660: e3sm.exe           00000000004A036F  component_mod_mp_         751  component_mod.F90
2660: e3sm.exe           000000000045D91B  cime_comp_mod_mp_        2891  cime_comp_mod.F90
2660: e3sm.exe           0000000000487D1C  MAIN__                    153  cime_driver.F90
2660: e3sm.exe           0000000000426122  Unknown               Unknown  Unknown
2660: libc-2.28.so       000015554DC8F7B3  __libc_start_main     Unknown  Unknown
2660: e3sm.exe           000000000042602E  Unknown               Unknown  Unknown
````

### LINKED ISSUE #5138: segfault with various DEBUG land cases
When I try to run a recent test added to nightly set (which is ERS) in DEBUG, I see segfaults on cori-haswell and cori-knl using either Intel or GNU.

```
SMS_D.hcru_hcru.I20TRGSWCNPRDCTCBC.cori-haswell_intel.elm-erosion
SMS_D.hcru_hcru.I20TRGSWCNPRDCTCBC.cori-haswell_gnu.elm-erosion

SMS_D.hcru_hcru.I20TRGSWCNPRDCTCBC.cori-knl_inel.elm-erosion
SMS_D.hcru_hcru.I20TRGSWCNPRDCTCBC.cori-knl_gnu.elm-erosion
```

I only get a stack trace with one of those -- cori-haswell and Intel.
The GNU cases do not give a stack trace, so assuming it is failing on same line.
Same for cori-knl/Intel -- fails without a stack.

```
165: forrtl: error (65): floating invalid
165: Image              PC                Routine            Line        Source
165: e3sm.exe           00000000070C9944  Unknown               Unknown  Unknown
165: e3sm.exe           00000000069F4620  Unknown               Unknown  Unknown
165: e3sm.exe           0000000000ECF35A  subgridavemod_mp_        1075  subgridAveMod.F90
165: e3sm.exe           0000000002B0EC27  cnpbudgetmod_mp_c        1148  CNPBudgetMod.F90
165: e3sm.exe           0000000002B0E20C  cnpbudgetmod_mp_c        1085  CNPBudgetMod.F90
165: e3sm.exe           000000000096E160  elm_driver_mp_elm         578  elm_driver.F90
165: e3sm.exe           00000000009097CB  lnd_comp_mct_mp_l         508  lnd_comp_mct.F90
165: e3sm.exe           000000000047C632  component_mod_mp_         751  component_mod.F90
165: e3sm.exe           0000000000439C2C  cime_comp_mod_mp_        2891  cime_comp_mod.F90
165: e3sm.exe           0000000000463FAC  MAIN__                    153  cime_driver.F90
165: e3sm.exe           0000000000402102  Unknown               Unknown  Unknown
165: e3sm.exe           000000000723AA8F  Unknown               Unknown  Unknown
```

```
if (carr(c) /= spval .and. scale_c2l(c) /= spval .and. scale_l2g(l) /= spval) then
```

Which is (test1 .and. test2 .and. test3)
It could be another situation where the original author assumed that if test1 is false, compiler would ignore test2 and test3, but I don't think we can assume that. Could be we just need more logic to only do certain tests when needed.

### PR DESCRIPTION
**Pull request by peterdschwartz**
Due to NaNs leaking through the initialization phase of ELM, ELM has not been able to 
adequately be run in DEBUG mode.  The first two commits allow all e3sm_land_developer 
tests to complete in DEBUG mode except for the one that fails in MOSART.  

Fixes #4820 
Fixes #5138 
Fixes #4946

[non-BFB]

### CODE REVIEW

**Comment by peterdschwartz** [comment]
@bishtgautam There was an unchecked divide-by-zero for ECA runs that I also fixed in PhosphorusStateUpdate3Mod that's also fixed in this PR.  ECA tests were still BFB.  

Also, I'll clean up the white space file differences that are left over from my testing.

**Comment by rljacob** [comment]
notes:  needs one more commit and finding how many bugs are closed by this.

**Comment by peterdschwartz** [comment]
Rebased the branch and went ahead and removed NaNs from missed data type files. 
It was BFB with the e3sm_land_developer test suite.
I ran all the e3sm_land_developer tests in DEBUG mode as well and the only FAILs were for betr and mosart but they seem unrelated to any floating point exceptions but due to un-initialized pointers.

**Comment by rljacob** [comment]
In the betr and mosart case it sounds like DEBUG is catching real issues which is what we want :)

File: components/elm/src/biogeochem/FireMod.F90 [hunk 1]
```diff
@@ -370,12 +371,14 @@ subroutine FireArea (bounds, &
 
               ! For non-crop -- natural vegetation and bare-soil
               if( veg_pp%itype(p)  <  nc3crop .and. cropf_col(c)  <  1.0_r8 )then
-                 if( .not. (btran2(p) .ne. btran2(p)))then !?shr_infnan_isnan(btran2(p))) then
-                    if (btran2(p)  <=  1._r8 ) then
-                       btran_col(c) = btran_col(c)+btran2(p)*veg_pp%wtcol(p)
-                       wtlf(c)      = wtlf(c)+veg_pp%wtcol(p)
+                 if(.not. shr_infnan_isnan(btran2(p)) ) then
```
Comment (bishtgautam) [inline components/elm/src/biogeochem/FireMod.F90:374]: Why would `btran2(p)` be `nan` to begin with?

File: components/elm/src/biogeochem/PhosphorusStateUpdate3Mod.F90 [hunk 2]
```diff
@@ -181,7 +181,7 @@ subroutine PhosphorusStateUpdate3(bounds,num_soilc, filter_soilc, num_soilp, fil
                             col_pf%sminp_to_plant_vr(c,j)*dt - col_pf%labilep_to_secondp_vr(c,j)*dt - &
                             col_pf%sminp_leached_vr(c,j)*dt ))
 
-                 if (temp_solutionp(c,j) < 0.0_r8) then
+                 if (temp_solutionp(c,j) < 0.0_r8 .and. (col_pf%labilep_to_secondp_vr(c,j)+col_pf%sminp_leached_vr(c,j) .ne. 0._r8) ) then
```
Comment (bishtgautam) [inline components/elm/src/biogeochem/PhosphorusStateUpdate3Mod.F90:184]: @qzhu-lbl if `col_pf%labilep_to_secondp_vr(c,j)+col_pf%sminp_leached_vr(c,j) = 0._r8`, the `col_pf%labilep_to_secondp_vr(c,j)` would be `nan`.  Is this code modification correct or would you recommend some other change?

File: components/elm/src/data_types/ColumnDataType.F90 [hunk 3]
```diff
@@ -9886,6 +9886,7 @@ subroutine col_pf_init(this, begc, endc)
            ptr_col=this%secondp_to_labilep)
 
     if ( nlevdecomp_full > 1 ) then
+       print *, ""
```
Comment (bishtgautam) [inline components/elm/src/data_types/ColumnDataType.F90:9889]: please delete this.

File: driver-mct/main/prep_rof_mod.F90 [hunk 4]
```diff
@@ -265,7 +265,7 @@ subroutine prep_rof_init(infodata, lnd_c2_rof, atm_c2_rof, ocn_c2_rof)
        call shr_sys_flush(logunit)
 
     end if
-
+
```
Comment (bishtgautam) [inline driver-mct/main/prep_rof_mod.F90:268]: please revert the edits (addition of wide spaces) in this file.

File: components/elm/src/biogeochem/FireMod.F90 (hunk 1 above)
Comment (peterdschwartz) [reply components/elm/src/biogeochem/FireMod.F90:374]: It's`nan` due to a finidat file used for a test that has it as `nan`.

File: components/elm/src/biogeochem/PhosphorusStateUpdate3Mod.F90 (hunk 2 above)
Comment (peterdschwartz) [reply components/elm/src/biogeochem/PhosphorusStateUpdate3Mod.F90:184]: @bishtgautam the master branch recently got a fix for this issue -- that's the one conflict for this PR it seems. So we could just opt for the change in master during the merge.

File: components/elm/src/biogeochem/FireMod.F90 (hunk 1 above)
Comment (bishtgautam) [reply components/elm/src/biogeochem/FireMod.F90:374]: How about fixing the finidat file of the test and leaving the code unchanged?

**Comment by rljacob** [comment]
status:  fixing some finidat files from some tests.

**Comment by ndkeen** [comment]
Should I wait on something to try this out and review?

**Comment by rljacob** [comment]
I think its ready to try out.

**Comment by peterdschwartz** [comment]
Yes it is ready @ndkeen

**Comment by ndkeen** [comment]
For these 2 tests (at least),
```
ERS_D.f19_g16.I1850GSWCNPRDCTCBC.pm-cpu_gnu.elm-ctc_f19_g16_I1850GSWCNPRDCTCBC
ERS_D.f19_g16.I20TRGSWCNPRDCTCBC.pm-cpu_gnu.elm-ctc_f19_g16_I20TRGSWCNPRDCTCBC
```

I see:
```
        wget failed with output:  and errput --2023-03-20 16:59:44--  https://web.lcrc.anl.gov/public/e3sm/inputdata/lnd/clm2/initdata_map/clmi.I1850GSWCNPECACNTBC.f19_g16.67a0f98.noNaN.0201-01-01-00000.nc
```

**Comment by peterdschwartz** [comment]
Added the 3 new finidat files to the server so hopefully that resolves the issue.

File: components/elm/src/biogeochem/FireMod.F90 [hunk 5]
```diff
@@ -371,14 +370,12 @@ subroutine FireArea (bounds, &
 
               ! For non-crop -- natural vegetation and bare-soil
               if( veg_pp%itype(p)  <  nc3crop .and. cropf_col(c)  <  1.0_r8 )then
-                 if(.not. shr_infnan_isnan(btran2(p)) ) then
                     if( btran2(p) .ne. spval) then
                        if (btran2(p)  <=  1._r8 ) then
                           btran_col(c) = btran_col(c)+btran2(p)*veg_pp%wtcol(p)
                           wtlf(c)      = wtlf(c)+veg_pp%wtcol(p)
                        end if
                     end if
```
Comment (bishtgautam) [inline components/elm/src/biogeochem/FireMod.F90]: After the removal of the `if` block, the code needs to be re-indented.

File: components/elm/cime_config/testdefs/testmods_dirs/elm/ctc_f19_g16_I1850GSWCNPRDCTCBC/user_nl_elm [hunk 6]
```diff
@@ -1,2 +1,2 @@
-finidat             = '$DIN_LOC_ROOT/lnd/clm2/initdata_map/clmi.I1850GSWCNPRDCTCBC.f19_g16.67a0f98.0401-01-01-00000.nc'
+finidat             = '$DIN_LOC_ROOT/lnd/clm2/initdata_map/clmi.I1850GSWCNPRDCTCBC.f19_g16.67a0f98.noNaN.0401-01-01-00000.nc'
```
Comment (bishtgautam) [inline components/elm/cime_config/testdefs/testmods_dirs/elm/ctc_f19_g16_I1850GSWCNPRDCTCBC/user_nl_elm:1]: How are these `finidat` files generated?

Comment (peterdschwartz) [reply components/elm/cime_config/testdefs/testmods_dirs/elm/ctc_f19_g16_I1850GSWCNPRDCTCBC/user_nl_elm:1]: There was only one variable that needed modification which was `btran2` so I edited the previous finidat file to replace all `nan` with the fill value instead using python netCDF4

**Comment by ndkeen** [comment]
I launched tests equivalent to changing all e3sm_integration tests to use DEBUG builds, and then adding the `invalid` check for GNU. I see 92 passes. So far the fails do not look to be related to use of a NaN and it would be good to make issues and start tracking those down.

So I think we should get this in and start using it. Perhaps could run some more basic tests

File: driver-mct/main/prep_rof_mod.F90 [hunk 7]
```diff
@@ -282,7 +282,7 @@ subroutine prep_rof_init(infodata, lnd_c2_rof, atm_c2_rof, ocn_c2_rof)
           call mct_aVect_zero(o2racc_ox(eoi))
        end do
        o2racc_ox_cnt = 0
-
+
```
Comment (bishtgautam) [inline driver-mct/main/prep_rof_mod.F90:285]: please undo this change.

File: driver-mct/main/prep_rof_mod.F90 [hunk 8]
```diff
@@ -576,7 +575,7 @@ subroutine prep_rof_merge(l2x_r, a2x_r, fractions_r, x2r_r, cime_model,o2x_r)
     character(*), parameter   :: subname = '(prep_rof_merge) '
 
     !-----------------------------------------------------------------------
-
+
```
Comment (bishtgautam) [inline driver-mct/main/prep_rof_mod.F90:578]: please undo this change.

File: components/elm/src/biogeochem/FireMod.F90 [hunk 9]
```diff
@@ -449,7 +449,7 @@ subroutine FireArea (bounds, &
            end if
         end do
      end do
-
+
```
Comment (bishtgautam) [inline components/elm/src/biogeochem/FireMod.F90:452]: please undo this change.

File: components/elm/src/biogeochem/PhosphorusDynamicsMod.F90 [hunk 10]
```diff
@@ -432,6 +432,7 @@ subroutine PhosphorusLeaching(bounds, num_soilc, filter_soilc, dt)
 
                ! limit the flux to a positive value
                sminp_leached_vr(c,j) = max(sminp_leached_vr(c,j), 0._r8)
+
```
Comment (bishtgautam) [inline components/elm/src/biogeochem/PhosphorusDynamicsMod.F90:435]: please undo this change.

**Review by bishtgautam — APPROVED** [review]
This is good to go and I have a few minor change requests to delete a few whitespaces.

**Comment by ndkeen** [comment]
What about files such as those below that set arrays to `nan`?

```
src/main/soilorder_varcon.F90
src/main/elm_interface_dataType.F90
src/utils/domainMod.F90
src/external_models/fates/biogeochem/EDPatchDynamicsMod.F90
src/external_models/emi/src/elm_stub/elm_types_stub/WaterfluxType.F90
src/external_models/sbetr/src/stub_clm/EcophysConType.F90
src/biogeochem/VOCEmissionMod.F90
src/dyn_subgrid/dynColumnStateUpdaterMod.F90
tools/mksurfdata_map/src/mkdomainMod.F90
```

**Review by ndkeen — APPROVED** [review]
I've run e3sm_integration and again with all tests DEBUG using compiler flag to check for `invalid` and I think it would be great to get these sort of changes merged in.

**Comment by peterdschwartz** [comment]
@bishtgautam New commit should take care of all the those whitespace changes.  

@ndkeen I can't change the files from submodules but I'll make a note to revisit those elm files soon.

**Pull request merged**